### ISSUE TITLE
iOS: Full screen reader mode

### ISSUE THREAD

**Issue by samuelclay** [samuelclay/NewsBlur#1137]
Like we have on Android, have an option that a tap on the story text (and not a link) causes the top header to recede for full screen reading.

**Comment by samuelclay** [comment]
Can we get the swipe on the left edge gesture to go back to story titles?

**Comment by samuelclay** [comment]
Also, tapping on images doesn't enter into full screen mode. I'd like to be able to tap anywhere that isn't a link.

**Comment by Dejal** [comment]
I think there was a reason for excluding images... perhaps a conflict with the long press menu.  But I'll take another look (next week).

**Comment by nriley** [comment]
Yeah, don’t want to lose the ability to view alt text on images.

**Comment by samuelclay** [comment]
This should just apply to a single tap.

**Comment by samuelclay** [comment]
Ok there’s two issues we need to address before we can ship this feature. One is that the text can’t move when going between full screen and regular. Second, and a bit less important, is that I want an interactive gesture to swipe back to the story list. 

I made a video showing how the nytimes app does this. 

https://share.icloud.com/photos/0iyKNjZz5iF1PuSKPo7BkTFfQ#Harvard_University

**Comment by Dejal** [comment]
The text moves to keep the relative scroll position the same... but I suppose I could add some padding at the top to offset it.

I'm not sure an interactive gesture is feasible in this situation, but I'll investigate.

**Comment by Dejal** [comment]
Done quite a bit of work to refine this, and really happy with the result. It now supports interactively hiding or showing the nav & status bars when scrolling up and down, like in the NY Times app, plus the text no longer moves when hiding/showing via tap. 

The edge swipe is still not interactive, though; I could work on that later if desired. I’ll leave this issue open for now; close it if you don’t want to bother with that.

**Comment by samuelclay** [comment]
There's some wonkiness when paging between stories that's caused by the header popping in and out.

**Comment by Dejal** [comment]
Not hiding the nav bar when at the top was intentional, to avoid leaving a blank space there (and avoid scrolling). If you'd prefer a blank space, I could change that, but it'll hide as you scroll up, so I don't see any benefit. Think of the tap as more a way to show the bar when hidden, than to hide it.

I'll work on the interactive edge swipe and cosmetic issues next week.

**Comment by nriley** [comment]
Agreed, I prefer the nav bar being visible versus having the content
move or a gray bar at the top.  (This is how the NYT and many other
iOS apps work, too.)

On Fri, Dec 14, 2018 at 07:08:02PM +0000, David Sinclair wrote:
> Not hiding the nav bar when at the top was intentional, to avoid leaving a blank space there (and avoid scrolling). If you'd prefer a blank space, I could change that, but it'll hide as you scroll up, so I don't see any benefit. Think of the tap as more a way to show the bar when hidden, than to hide it.
> 
> I'll work on the interactive edge swipe and cosmetic issues next week.
> 

-- 
--Nicholas

**Comment by samuelclay** [comment]
I think hiding the nav bar while at top is preferable if we can get interactive edge swipe working. Once I'm in full screen mode, I want to a) know that I'm in it and b) stay in it.

**Comment by Dejal** [comment]
Interactive edge swiping is a separate problem (and does work when the nav bar is visible). 

I'm a bit confused; I initially had it only hide and show the nav bar via tap, but you posted a video saying you prefer the way the NYT app does it, showing the nav bar hiding and showing when scrolling, which is actually a standard iOS convention. So I spent time to implement that, which does use the system mechanism. I think that showing the bar when quickly scrolling upwards or via a tap is best.

I could remove the logic to show the bar when scrolled to the top (I actually initially had it show when you reach the bottom too, to make it easier to navigate or access the menu, but decided to only do that at the top). But I don't think we should disable the system show/hide swipe gesture.

I think full screen reading should be used interactively when reading, not a permanent mode that has to be toggled.

**Comment by samuelclay** [comment]
> I think full screen reading should be used interactively when reading, not a permanent mode that has to be toggled.

I think it should be toggled per session. When you go back to story title and back into the story detail, you should have to tap back into full screen mode. But once in, it's full screen for all stories.

**Comment by Dejal** [comment]
So to be clear, do you want me to disable showing and hiding the nav bar when scrolling, like in the NYT app?  Currently, when full-screen is enabled in the prefs, you go into a story, and the bar is shown, then as you scroll down the bar will be interactively hidden, and will remain so as you keep scrolling, but if you scroll back up it will be interactively shown, and is forced shown when you reach the top (e.g. if you scrolled slowly enough to not show it via the gesture; try scrolling slowly; it's really nice).  Plus you can show or hide it with a tap, except when at the top.

So, it sounds like you don't like the interactive nav bar gesture after all, and don't like showing it at the top. You just want it to hide or show the nav bar on tap, and not via scrolling?

I personally think that'd be a worse experience, but if that's what you prefer, that's what I'll do. I just want to make sure I understand what you want.

(Another option could be to offer both options, so the preference has options for no full-screen, only via tap, and also interactively via scrolling.)

**Comment by samuelclay** [comment]
The new behavior that is apparent with slow scrolling makes the text move by at different speeds. This is not what I had in mind. I think when a full screen mode is enabled, just go right into it by hiding the top toolbar and ensuring the text doesn't move. The text is first-class, everything should cater to it, which means it should never move at a different speed or be moved at all. 

I'm not sure that extra options are best. I'd prefer to keep it at a single option as it is.

**Comment by Dejal** [comment]
It's a system-standard behavior, but I understand your priorities.  I will disable the scroll hiding gesture, and allow hiding at the top, without moving the text.

**Comment by samuelclay** [comment]
Here's an idea for the interactive swipe. How about as soon as a press-down gesture is detected in the left 20% of the screen, show the toolbar and then interactive swipe can begin. I've played with this for a bit and I feel that because the text doesn't move, we can show and hide the toolbar at will.

**Comment by Dejal** [comment]
I could try that. I'm not sure it'd work, as the edge touch may block the swipe gesture (since the finger isn't lifted, it'd treat it as the same gesture sequence), but it may be feasible.

I did some work on hiding the bar at the top last night; leaving empty space is okay, if a little ugly, but as soon as you scroll it jumps a little, since it has over-scrolled to expose that space (and keep the text in the same place). I don't think you'd like that, so I'll experiment with always adding more space and under-scrolling when the bar is shown, to see if that works better.

**Comment by Dejal** [comment]
I'm struggling a bit to have a satisfactory experience with this. I've used up my time for the month, so will come back to this next month, hopefully with some fresh insights, but if either of you two want a fun little challenge for holiday time off, you're welcome to have a play with it.  I've committed where I'm at (on the Dejal branch), but note that it isn't release-ready.  (Though if a release were desired, you could just hide the full-screen preference.)

Current issues:

1. When at the top, hiding the nav & status bars now also uses JavaScript to add a margin to the top of the page, but this can sometimes result in a very quick "blink" if the page content and scrolling doesn't occur at the same instant. Another solution could be to always have a margin, and scroll a little to hide it when not full-screen. Or don't worry about the margin, and accept the previously mentioned slight scroll jump. Or go back to showing the bars at the top.
2. Unwanted space can occur when hiding the bars for one story, swiping to the next story, and showing the bars. It fixes itself when you go to the next story, but shouldn't do that. Almost certainly because the next story is already loaded when updating the current story, so the fix will probably involve updating them, but I haven't spent too much time on that yet.
3. When the bars are hidden, starting an edge swipe will now show the bars, but then you have to swipe again to go back to the stories list. As predicted above, it blocks the gesture. This could probably be improved by adding logic at the `EventWindow` level to detect the edge touch before the gesture is recognized.

I'm sorry this is taking a while. As I said, I'd be happy for one of you to tackle it if you like, otherwise I'll come back to it next month.

**Comment by samuelclay** [comment]
Download the Happy Cow app. I like how it scrolls, as the top nav bar doesn't go away until the top of the content is past it. We don't need to full screen immediately on tap if the content isn't at the top. We can simplify the experience, it doesn't have to be all inclusive of every scenario. 

There's some bugs in the current implementation around spacing and it's gotten to the point where I think we might need to stat from scratch on the feature, rethinking when it deploys.

**Comment by Dejal** [comment]
I've just bought the Happy Cow app, as suggested.  It scrolls with what looks like a custom variation of the standard dynamic hiding mechanism.  As you say, it doesn't hide the bars until some distance from the top, avoiding the quick hide of the standard mechanism that you didn't like.

Other than that, it seems pretty standard; it still has dynamic hiding when fast scrolling: a fast scroll downward will hide the bar, a fast scroll upward will show it, and slowly scrolling to the top also shows it, just like I implemented.  I'm getting mixed messages; you previously said you didn't want it hiding or showing except on tap. Have you changed your mind about that?

I still prefer the dynamic hiding and showing at the top, and it would solve the issue of managing empty space at the top. So to avoid the issue you had with it, I could disable dynamic hiding when at the top, and enable it once scrolled down a bit.

The unwanted space and edge swipe are separate issues.

Please let me know if you want me to re-add dynamic hiding, with the top delay as suggested. (Fortunately I just commented out the code, so it won't be hard to re-enable.)  Or if you'd prefer to remove this feature entirely for now, and come back to it in the future.  (As tempting as that may be, it would involve more time getting back up to speed with the issues.)

**Comment by samuelclay** [comment]
Let's try disabling dynamic hiding while at top, but there are some other UI issues that I'm afraid might not be able to be sorted out without a bigger rewrite. I'm willing to hit pause on this feature, save for trying that one change.

**Comment by samuelclay** [comment]
Oh this is excellent. Really nice work here. Ok, one addition. When I tap on a story that's already full screen, also show the story traversal buttons at the bottom when the toolbar on top shows up. And tapping again should hide them if the toolbar also hides.

**Comment by Dejal** [comment]
Done! Closing this issue now; feel free to re-open if you want any further tweaks.

**Comment by samuelclay** [comment]
Ok I want to change something. I want to get rid of the tap gesture. If you have full screen mode enabled it turns on for every story just as it does now. But you shouldn’t be able to get the toolbar back since swiping a bit to the left or right gets you the story traversa buttons back. 

Also can we add a segmented control to story detail view menu: [full screen | tool bar ]

**Comment by samuelclay** [comment]
Now that we've disabled the tap to toggle full screen, can we add a pinch gesture to show/hide the toolbar?

**Comment by samuelclay** [comment]
Now that full screen is not enabled by a tap, can we get a more responsive swipe back to the story list?

Second, I’m fine with the form sheet presentation of the safari view on iPad but it should remain a push view on iPhone. Otherwise I can’t swipe to go back to the story detail.

**Comment by nriley** [comment]
Since I didn't actually see the results of this until recently, I didn't quite understand the above conversation when I read it. What are the advantages from using the form sheet presentation for in-app Safari on iPad? From my standpoint it just wastes a substantial portion of the display...

**Comment by samuelclay** [comment]
Isn't this ticket about the automatic hiding of the toolbar while in the story detail view? Your comment seems like it has to do with the in-app browser on iPad, which I agree I'm not in love with the new form sheet presentation.

**Comment by nriley** [comment]
Yup! I just attached it here because the relevant commit is attached to this ticket. Happy to file another if that would be easier.

**Comment by Dejal** [comment]
I believe the form sheet presentation was to work around some issue, though I forget the details at present.

Let's make a new issue for that.  I may be able to come up with a better solution.
